add ltc2358 support

kseerp opened the pull request:
This PR enables the integration of the DC2677A board with the SoC FPGA Cyclone 5 FPGA. 
The setup involves utilizing the [axi-dc2677a-hdl-project](https://github.com/analogdevicesinc/hdl/pull/1130) repository for the required HDL project.

kseerp commented:
update: fix bindings and addressed checkpatch error

kseerp commented:
v2
- `__ltc2358_set_sampling_freq` return directly after function call
- prioritize hardware read over reading cached values

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,131 @@
+      reg:
+        description: Channel number.
+        minimum: 0
+        maximum: 7
```
If this is going to be the only property on the channel node, please drop the node. I understand you're using this to instantiate IIO channels but you can just have all of them exposed to userspace. Sometimes might happens that an input can be hot plugged in some design containing this HW.

Some other times, yes, might make sense to have these channel nodes that will "decide" if a channel is exposed via IIO or not. But in those cases you should have other "mandatory" properties that really define the input that you want in a particular channel that make it impossible to do any "hot plug".

kseerp replied:
I guess I have to retain this node since there is a need to make the input range as dt property, in order to assign the input range property to a particular channel.

nunojsa replied:
yes, you need it...

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,131 @@
+  - pwms
+  - pwm-names
+  - pd-gpios
+  - csn-gpios
```
are the above really mandatory properties?

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,131 @@
+  - pd-gpios
+  - csn-gpios
+
+unevaluatedProperties: false
```
I think you want `additionalProperties` in here

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,131 @@
+
+      adc@0 {
+          compatible = "adi,ltc2358-18";
+          reg = <0>;
```
this is a platform device with an IOMAP... This reg property does not look right

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,131 @@
+  - |
+    dev {
+      #address-cells = <1>;
+      #size-cells = <0>;
```
this also likely wrong...

kseerp replied:
I'm not yet familiar with setting bindings for axi or platform devices. What I'm used to is working with SPI. Should I follow the same bindings in this case, or do I need to replicate what I already have in my dts?

nunojsa replied:
You need to replicate a proper configuration which `reg = <0>;` is not :)... Being a platform device I'm not sure the `dev {` part is needed but I guess it does not arm

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+#include <linux/kernel.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/of.h>
```
why?

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+
+#include <linux/iio/iio.h>
+#include <linux/iio/types.h>
+#include <linux/iio/sysfs.h>
```
is this one really needed? Make sure to include only the headers you need....

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+	csn_gpio = devm_gpiod_get_optional(&pdev->dev, "csn", GPIOD_OUT_LOW);
+	if (IS_ERR(csn_gpio))
+		return dev_err_probe(&pdev->dev, PTR_ERR(csn_gpio),
+				     "Failed to get CSN GPIO\n");
```
You are not doing a thing with the above gpios... So, either fully implement them or remove the above lines...

kseerp replied:
Yeah, I would have to remove the gpios. We could set the csn and pd gpios to low by default on the hdl part.

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+			return dev_err_probe(&pdev->dev, -EINVAL,
+					     "Invalid vref voltage\n");
+
+		st->vref_mv = (ret * 2) / 1000;
```
nit: I think you don't really need the parenthesis...

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+	if (IS_ERR(st->regs))
+		return PTR_ERR(st->regs);
+
+	platform_set_drvdata(pdev, indio_dev);
```
I'm not seeing `platform_get_drvdata()` being called so you can drop the above

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+
+	mem = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	st->regs_size = resource_size(mem);
+	st->regs = devm_ioremap_resource(&pdev->dev, mem);
```
`devm_platform_ioremap_resource()`

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+	ltc2358_axiadc_write(st, ADI_AXI_REG_RSTN, 0);
+	mdelay(10);
+	ltc2358_axiadc_write(st, ADI_AXI_REG_RSTN, ADI_AXI_REG_RSTN_MMCM_RSTN);
+	mdelay(10);
```
use `fsleep()`

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+{
+	struct regulator *reg = data;
+
+	regulator_disable(reg);
```
just call `regulator_disable(data)`. You can also rename data -> reg

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+		ltc2358_axiadc_write(st, AXI_SOFTSPAN_ADDR(reg),
+					LTC2358_0V_1_5V_VREF);
+
+		ltc2358_axiadc_write(st, AXI_REG_CHAN_CNTRL(reg), 1);
```
just do the above two lines for all the channels and remove the `device_for_each_child_node()` loop

kseerp replied:
Can I still keep the `device_for_each_child_node()` loop to configure an input range for a specific channel, based on what is set in the device tree? Same as your implementation [https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/iio/dac/adi,ltc2688.yaml#L137](ltc2688)

nunojsa replied:
Yes, you should... I'm not sure if it's needed or not (it looks it is from the above code) but after going though `device_for_each_child_node()` you need to loop over the channels that were not configured via dts to st the default input value as you're doing now.

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+	IIO_ENUM_AVAILABLE_SHARED("input_range", IIO_SHARED_BY_ALL,
+						&ltc2358_softspan),
+	{ },
+};
```
Drop this and make it a devicetree property... Typically this kind of scale is something that you don't really change at runtime. I had something similar for a [DAC](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/iio/dac/adi,ltc2688.yaml#L65) and also made the question upstream (as this parameter can be completely controlled in SW) and I was told to just go with a dts parameter.

kseerp replied:
Regarding this, is it acceptable to use the softspan code instead of setting a specific input range value in the dt property. The input range changes depending on the Vref, so could I simply use the softspan code for the property value? For example, (adi,softspan-code = <1>;, adi,input-range-number = <1>;, or adi,config-number = <1>;)

I see the idea of this implementation here: [linduino-ltc](https://github.com/analogdevicesinc/Linduino/blob/bff9185178d2bf694d0fed14a85392f21655c7de/LTSketchbook/Part%20Number/2000/2300/2348/DC2094A/DC2094A.ino#L266-L285)

nunojsa replied:
Hmm not sure... Ideally you would have a volts based property but since you cannot really know the proper values, the code seems to be the only option available. But make sure to document each code (dependent on vref) in the description.

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+
+static const struct attribute_group ltc2358_attribute_group = {
+	.attrs = ltc2358_attributes,
+};
```
Also drop this. To me this looks more like a debug thing. So make it a debugfs property

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,555 @@
+		*val2 = st->chip->resolution;
+
+		return IIO_VAL_FRACTIONAL_LOG2;
+
```
...

kseerp replied:
Sorry, this is my first time encountering this comment, and I have no idea what it means. Should I include additional code or remove the line after the `return IIO_VAL_FRACTIONAL_LOG2` statement?"

kseerp commented:
v3 

bindings:
- added adi,softspan-code property and documentation 
- created proper configuration on example dt
- replaced unevaluatedProperties with additionalProperties

ltc2358:

- removed unused headers
- removed samp_freq field in struct
- replaced dev_err_probe with dev_err in ltc2358_set_sampling_freq function
- assigned struct device *dev with &st->pdev->dev to make code neater
- removed io attribute and added debugfs attribute
- changed call to regulator_disable(data)
- replaced axiadc_reset mdelay() to fsleep()
- removed csn and pd gpios, set active low by default on hdl part
- replaced devm_ioremap_resource() to devm_platform_ioremap_resource()
- retained instantiation of num_channels in channel_config function based on the number of child nodes in the device to take into account that the sampling frequency can vary depending on the number of channels utilized
- misc changes(removed parenthesis, lines...)

tested current version DC2677A board with socfpga cyclone 5 board

nunojsa reviewed (changes requested):
So, almost there...

One thing that I completely forgot to mention is that these kind designs are done using the backend hdl driver [cf_axi_adc_core.c](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/cf_axi_adc_core.c).


You're not duplicating that much code but still I would prefer for you to use the typical way of supporting these device. You would still have most of the driver as you have now...  You would just need to properly setup `struct axiadc_converter` and remove the IIO stuff as the IIO device is created in the cf_axi_adc_core.c drive.

One advantage of the above is that you can use the generic osc ADC/DAC plugins for capturing data in osc.

kseerp commented:
> One thing that I completely forgot to mention is that these kind designs are done using the backend hdl driver [cf_axi_adc_core.c](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/cf_axi_adc_core.c).

I'm using an axi register map that does not seem to be included in the file. [SOFTSPAN](https://github.com/analogdevicesinc/hdl/blob/master/docs/regmap/adi_regmap_adc.txt#L971-L976) and [REG_CHAN_RAW_DATA](https://github.com/analogdevicesinc/hdl/blob/master/docs/regmap/adi_regmap_adc.txt#L772-L776)

Would it be okay to open a PR to have it included?

nunojsa commented:
> Would it be okay to open a PR to have it included?

Sure, I think that makes sense...

kseerp commented:
v4:

- extended ltc235x family support
- utilized cf_axi_adc_core and removed iio instantiation
- added softspan and channel raw data register map in cf_axi_adc header

kseerp commented:
v5:

- removed unused headers
- misc changes(removed extra tabs)

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,433 @@
+	if (ret)
+		return ret;
+
+	mutex_lock(&st->lock);
```
This seems to be the only place where the lock is used so I don't think you need it... `iio_device_claim_direct_mode()` already serializes accesses to `__ltc235x_set_sampling_freq()` on top of making sure the sampling frequency is not changed if buffering is on-going.

kseerp commented on Documentation/devicetree/bindings/iio/adc/adi,ltc2358.yaml:
```diff
@@ -0,0 +1,157 @@
+
+    required:
+      - reg
+      - adi,softspan-code
```
Apologies for forgetting to respond to your previous comment. I made "adi,softspan-code" a required property for I replicated the way this [driver](https://github.com/analogdevicesinc/Linduino/blob/bff9185178d2bf694d0fed14a85392f21655c7de/LTSketchbook/Part%20Number/2000/2300/2348/DC2094A/DC2094A.ino#L342) was implemented, and I feel it should be set as mandatory based on what I also see in the datasheet. If this approach may not be ideal, I could consider implementing a default value for softspan in the code.

nunojsa replied:
Doesn't the device has a default softspan code?

> and I feel it should be set as mandatory based on what I also see in the datasheet.

Why do you feel that way? I'm not against having it as mandatory at all as long as it makes sense :)

kseerp replied:
My mistake on this one. I have confirmed that the hdl part has already set a default value for the softspan, which is 7. In this case, we could make the 'adi,softspan-code' property, as well as the 'reg' property, not mandatory. Would this be valid?

nunojsa replied:
The reg property needs to be mandatory... It is a way for you to identify some channel. Objectively speaking, since you only have one configuration property (softspan), it would not make sense to have a channel node without setting up that property but in terms of correctness (and extendability), yeah, let's make softspan as a optional property.

> which is 7

You cal also set that in the bindings with `default: 7`.

kseerp commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,433 @@
+	ret = of_parse_phandle_with_args(np, "spibus-connected", NULL, 0, &args);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to parse spibus-connected\n");
+
```
I noticed that only the properties within the parent nodes are being detected. It appears that properties in child nodes are not being parsed correctly, that's why I implemented the code you mentioned in order to properly parse the properties inside the child nodes like reg and softspan-code. Are there any workarounds for this one?

nunojsa replied:
If you store your spi device in your main struct then you can easily get it's FW properties... And you can get your main struct by using `conv->phy`.  And actually I'm not sure why we have the FW parsing in the post_setup() hook. You can very well do it in the probe() function. You would need to cache the values and then write them in here.

Anyways, even if you opt for keeping this as-is, you can read your properties without the hack you are using now...

kseerp commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,433 @@
+		return -ENOMEM;
+
+	st = iio_priv(indio_dev);
+	st->spi = spi;
```
Yeah, I overlooked on this one.

kseerp replied:
I would like to clarify this one, it seems that this is required to run the driver, as I also found the same implementation in another driver that also utilized the cf_axi_adc https://github.com/analogdevicesinc/linux/blob/daad63c8ec30312d90beba4db9dfceac5a3e67c3/drivers/iio/adc/adaq8092.c#L360-L362

nunojsa replied:
I'm failing to understand why it's needed? To me, it looks like we are using `devm_iio_device_alloc()` just to allocate our internal struct which is obviously a wast of memory.

> as I also found the same implementation in another driver that also utilized the cf_axi_adc

It does not make it right :). I'm no saying is wrong but I need to be proved otherwise

kseerp commented:
v6:

- added all ltc235x compatibility in the bindings
- created separate function for device tree parsing
- made adi,softspan-code an optional property and set 7 as the default
- removed mutex_lock
- utilized the device_property API for property parsing

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,444 @@
+			fwnode_handle_put(child);
+			return dev_err_probe(dev, -EINVAL,
+					     "Missing channel number\n");
+		}
```
Rework this please.... Typically, the style in linux is to first look for errors. So this could be:

```
if (ret) {
    fwnode_handle_put(child);
    return dev_err_probe(dev, -EINVAL, "Missing channel number\n");
}

if (reg >= st->num_channels) {
    fwnode_handle_put(child);
    return dev_err_probe(dev, -EINVAL,"Invalid channel number\n");
}
```

Looks better than the if() else, right?

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,444 @@
+				}
+			} else {
+				st->softspan[reg] = 7;
+			}
```
The indentation is wrong. You can also rewrite this like:

```
st->softspan[reg] = 7;
ret = fwnode_property_read_u32(child, "adi,softspan-code", &st->softspan[reg]);
if (!ret) {
    ...
}
```

Note that `st->softspan[reg]` is only modified if the property does exist... So you can just put the default value before. That makes the code a bit neater.

nunojsa commented on drivers/iio/adc/ltc2358.c:
```diff
@@ -0,0 +1,444 @@
+
+	indio_dev = devm_iio_device_alloc(dev, sizeof(*st));
+	if (!indio_dev)
+		return -ENOMEM;
```
I'm still not convinced the above is needed... Can you just:

```
st = devm_kzalloc(dev, sizeof(*st), GFP_KERNEL);
```

and see if it still works?

kseerp replied:
It appears to be working as expected.

kseerp commented:
v7:

updated license specification
reworked device tree parsing
changed memory allocation for struct
